core/src/main/scala/com/eharmony/spotz/optimizer/grid/Grid.scala

```diff
@@ -0,0 +1,105 @@
+package com.eharmony.spotz.optimizer.grid
+
+import com.eharmony.spotz.util.Logging
+
+import scala.collection.mutable.ArrayBuffer
+import scala.language.postfixOps
+
+/**
+  * This grid implementation computes hyper parameter values using the algorithm found here:
+  *
+  * <link>http://phrogz.net/lazy-cartesian-product</code>
+  *
+  * The entire cartesian product space of the grid is not ballooned into memory, but is
+  * instead computed on demand through the apply method.
+  *
+  * Grid elements are accessible through an index, similarly to an IndexedSeq.
+  *
+  * {{{
+  *   import com.eharmony.spotz.Preamble._
+  *
+  *   val grid = new Grid[Point](Map(
+  *     ("x1", Range.Double(0.0, 1.0, 0.1)),
+  *     ("x2", Range.Double(0.0, 1.0, 0.1))
+  *   )
+  *
+  *   val length = grid.length            // 100
+  *   val firstElement = grid(0)          // Point(Map(x1 -> 0.0, x2 -> 0.0))
+  *   val lastElement = grid(99)          // Point(Map(x1 -> 1.0, x2 -> 1.0))
+  *   val lastElementPlusOne = grid(100)  // IndexOutOfBoundsException
+  * }}}
+  *
+  * The factory function defines the transformation from the Map of sampled hyper parameters
+  * to the point P, which is passed implicitly.
+  *
+  * The <code>GridSearch</code> algorithm will use this class to iterate through all the grid element
+  * points by index and pass them for evaluation to the objective function.
+  *
+  * @author vsuthichai
+  */
+class Grid[P](
+    gridParams: Map[String, Iterable[_]])
+    (implicit factory: (Map[String, _]) => P)
+  extends Serializable
+  with Logging {
+
+  assert(gridParams.nonEmpty, "No grid parameters have been specified")
+
+  private lazy val gridRows = computeGrid()
+  lazy val length = computeLength()
+  lazy val size = length
```

Comment:
There are types that seem to support both so I did the same.

List, IndexedSeq to name a few.  Didn't feel it hurt either way, I think it's a matter of preference which to use since they both convey the same idea.